src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts

```diff
@@ -0,0 +1,911 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import { Disposable } from 'vs/base/common/lifecycle';
+import { localize } from 'vs/nls';
+import { ICommandService } from 'vs/platform/commands/common/commands';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { IDialogService } from 'vs/platform/dialogs/common/dialogs';
+import { IStorageService, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { IChatCommandExecutionService } from 'vs/workbench/contrib/chat/browser/chat';
+import { isCommandButtonRenderData } from 'vs/workbench/contrib/chat/browser/chatListRenderer';
+import { IChatResponseViewModel } from 'vs/workbench/contrib/chat/common/chatViewModel';
+
+const GRANTED_RUN_COMMAND_PERMISSION_KEY = 'chat.grantedRunCommandPermission';
+
+export class ChatCommandExecutionService extends Disposable implements IChatCommandExecutionService {
+	declare readonly _serviceBrand: undefined;
+	constructor(
+		@ICommandService private readonly _commandService: ICommandService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IStorageService private readonly _storageService: IStorageService,
+		@IDialogService private readonly _dialogService: IDialogService
+	) {
+		super();
+	}
+	async acceptResponse(response: IChatResponseViewModel | string | undefined): Promise<void> {
+		if (!this._configurationService.getValue('github.copilot.experimental.runBestCommandMatch')) {
+			return;
+		}
+		if (typeof response === 'string') {
+			return;
+		}
+		const button = response?.renderData?.renderedParts.find(part => isCommandButtonRenderData(part));
+		if (!button || !('command' in button) || !commandsToRun.includes(button.command.id)) {
```

Comment:
I think this should only be enabled for the `@vscode` participant or participants that opt-in to this, which they could do by setting another flag when they stream back the button. It makes sense when you are asking it to recommend a command, and might not make sense at all in a different scenario.

Reply:
I have hardcoded it for the `vscode` agent for now. Do you mean adding a property to `IResponse` like `allowAutomaticRunCommands`?